File: straccia-camicia.py
import random
import itertools

class Mazzo(list):
    def __init__(self, *args):
        list.__init__(self, *args)
        self.semi = ['B','C','D','S']
        self.numeri = [str(i) for i in range(1,11)]

        for s in self.semi:
            for n in self.numeri:
                self.append(s+n)

    def mescola(self):
        random.shuffle(self)
        
    def __str__(self):
        s = 'Mazzo:'
        for c in self:
            s+= ' '+c
        return s

class Giocatore():
    def __init__(self, mano, nome):
        self.mano = mano
        self.nome = nome
        
    def gioca_carta(self):
        return self.mano.pop()
        
    def __str__(self):
        s = 'Mano '+self.nome+':'
        for c in self.mano:
            s += ' '+c
        return s

class Piatto(list):
    def __init__(self, *args):
        list.__init__(self, *args)
        self.chiede = 0

class Gioco():
    def __init__(self, mazzo = Mazzo()):
        self.mazzo = mazzo
        self.mazzo.mescola()
        self.g1 = Giocatore(self.mazzo[:20], 'g1')
        self.g2 = Giocatore(self.mazzo[20:], 'g2')
        self.piatto = Piatto()
        self.g_attivo = self.g1
        self.g_inattivo = self.g2
        self.numero_mossa = 0
        self.fine = False
    
    def valori_iniziali(self):
        print(self.mazzo)
        print()
        print(self.g1)
        print()
        print(self.g2)
        print()
        
    def scambia_giocatore(self):
        if self.g_attivo == self.g1:
            self.g_attivo = self.g2
            self.g_inattivo = self.g1
        elif self.g_attivo == self.g2:
            self.g_attivo = self.g1
            self.g_inattivo = self.g2
    
    def consegna_piatto(self):
        self.g_inattivo.mano = self.piatto + self.g_inattivo.mano
        self.piatto = Piatto()

    def gioca_mano(self):
        self.numero_mossa += 1
        if len(self.g_attivo.mano) != 0:
            carta = self.g_attivo.gioca_carta()
            self.piatto.append(carta)
            #print(str(self.numero_mossa)+' - '+self.g_attivo.nome+': '+carta)
            if carta[-1] in ['1','2','3']:
                self.piatto.chiede = int(carta[-1])
                self.scambia_giocatore()
            else:
                if self.piatto.chiede == 0:
                    self.scambia_giocatore()
                elif self.piatto.chiede == 1:
                    self.consegna_piatto()
                    self.scambia_giocatore()
                else:
                    self.piatto.chiede -= 1                
        else:
            self.consegna_piatto()
            #print(str(self.numero_mossa)+' - '+self.g_attivo.nome+': perde')
            self.fine = True                


mosse = 0
i = 0
while True:
    i += 1
    g = Gioco()
    while not g.fine:
        g.gioca_mano()
    if g.numero_mossa > mosse:
        mosse = g.numero_mossa
        print('\n'+str(i)+' '+str(mosse)+' '+str(g.mazzo))

# 2009 C5 C9 B3 C10 B8 B9 D10 D6 S1 D8 D3 D5 S5 B6 C7 S2 D2 B4 C6 S10 B7 D4 S6 S7 D7 B5 C3 D9 C8 S9 B1 S8 C2 C1 D1 S3 C4 B10 S4 B2

File: straccia_camicia.cpp
#include <vector>
#include <iostream>     // std::cout
#include <algorithm>    // std::shuffle
#include <array>        // std::array
#include <random>       // std::default_random_engine
#include <chrono>       // std::chrono::system_clock
#include <string>
#include <list>

using namespace std;

class Mazzo {
public:
    vector<int> mazzo = {1,2,3,1,2,3,1,2,3,1,2,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
    
    void mescola() {
        unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
        shuffle(mazzo.begin(), mazzo.end(), std::default_random_engine(seed));
    }
};

class Giocatore {
public: 
    list<int> mano;
    string nome;
    
    Giocatore(string nome, list<int>* mano) : nome(nome), mano(*mano) {}
    
    int gioca_carta() {
        int ret = mano.back();
        mano.pop_back();
        return ret;
    }
    
};

class Piatto {
public:
    vector<int> piatto = {};
    int chiede = 0;
};

class Gioco {
public:
    Mazzo mazzo;
    Piatto piatto;
    Giocatore* g1,* g2;
    Giocatore* giocatori[2];
    int numero_mossa = 0;
    bool fine = false;
    int piatto_chiede = 0;
    bool log;
    
    
    Gioco(bool mescola, bool log) : log(log) {
        if (mescola) { 
            mazzo.mescola(); 
        }
        g1 = new Giocatore(string("Marco"), new list<int>(mazzo.mazzo.begin(), mazzo.mazzo.begin() + 20));
        g2 = new Giocatore(string("Luca"), new list<int>(mazzo.mazzo.begin() + 20, mazzo.mazzo.end()));
        giocatori[0] = g1;
        giocatori[1] = g2;
    }
    
    void valori_iniziali(){
        cout << "Mazzo:\n";
        for (int i : mazzo.mazzo) {
            cout << i << " ";
        }
        cout << endl;
        cout << "Mano giocatore 1:\n";
        for (int i : g1->mano) {
            cout << i << " ";
        }
        cout << endl;
        cout << "Mano giocatore 2:\n";
        for (int i : g2->mano) {
            cout << i << " ";
        }
        
    }
    
    void scambia_giocatori() {
        Giocatore* tmp = giocatori[0];
        giocatori[0] = giocatori[1];
        giocatori[1] = tmp;
        cout << "giocatori scambiati" << endl;
    }
    
    void consegna_piatto() {
        reverse(piatto.piatto.begin(), piatto.piatto.end());
        for (int i : piatto.piatto) {
            giocatori[1]->mano.push_front(i);
        }
        piatto = *(new Piatto());
    }
    
    void gioca_mano() {
        numero_mossa++;
        int carta = 0;
        if(giocatori[0]->mano.size() != 0){
            carta = giocatori[0]->gioca_carta();
            piatto.piatto.push_back(carta);
            if (log) { 
                cout << numero_mossa << " " << giocatori[0]->nome << " " << carta << endl; 
            }
            if (carta > 0) {
                piatto_chiede = carta;
                scambia_giocatori();
            } else {
                if (piatto_chiede == 0) {
                    scambia_giocatori();
                } else if (piatto_chiede == 1){
                    consegna_piatto();
                    scambia_giocatori();
                } else {
                    piatto_chiede--;
                }
            }
        } else {
            consegna_piatto();
            if (log) {
                cout << numero_mossa << " " << giocatori[0]->nome << "perde";
            }
            fine = true;
        }
    }
    
    void gioca(){
        while(!fine) {
            gioca_mano();
        }
    }
};

int main(){
    int mosse = 0;
    int max_iterazione = 10;
    Mazzo mazzo_base;
    Gioco g(true, true);
    g.gioca();
    
}